### DISCUSSION TITLE
Setting time_base when adding a stream to an output container

### DISCUSSION THREAD

**Discussion by daveisfera** [PyAV-Org/PyAV#1525] [4. Bugs]
## Overview
Accessing `time_base` now raises an error with 13.0 when specifying `time_base` when calling `add_stream` on an output container

## Expected behavior
Able to specify the `time_base` on a stream like previous versions

## Actual behavior
`RuntimeError` raised when it's called

Traceback:
```
Traceback (most recent call last):
  File \"/usr/src/app/test.py\", line 129, in test
    main()
  File \"/usr/src/app/test.py\", line 81, in _test
    stream = container.add_stream(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File \"av/container/output.pyx\", line 132, in av.container.output.OutputContainer.add_stream
  File \"av/stream.pyx\", line 111, in av.stream.Stream.__setattr__
  File \"av/codec/context.pyx\", line 511, in av.codec.context.CodecContext.time_base.__set__
RuntimeError: Cannot access 'time_base' as a decoder"
```

## Investigation
Caused by [this commit](https://github.com/daveisfera/PyAV/commit/ba1ace7d0f5025ad10ed92e3f6dc85dca6374fa5)

## Reproduction
Call `add_stream` with `time_base`, like this:
```
container.add_stream("h264", time_base=Fraction(1, 1000))
```

## Versions

- OS: Debian Bookworm
- PyAV runtime: 13.0
```
PyAV v13.0.0
library configuration: --prefix=/usr --enable-shared --disable-static --disable-debug --disable-doc --disable-ffplay --enable-gpl --enable-libfdk_aac --enable-libx264 --enable-libx265 --enable-librsvg --enable-openssl --enable-nonfree --enable-libaom --extra-libs=-lpthread
library license: nonfree and unredistributable
libavcodec     60. 31.102
libavdevice    60.  3.100
libavfilter     9. 12.100
libavformat    60. 16.100
libavutil      58. 29.100
libswresample   4. 12.100
libswscale      7.  5.100
```
- PyAV build:
```
pip3 install av==13.0.0 --no-binary av
```

## Research

I have done the following:

- [X] Checked the [PyAV documentation](https://pyav.basswood-io.com)
- [X] Searched on [Google](https://www.google.com/search?q=pyav+how+do+I+foo)
- [X] Searched on [Stack Overflow](https://stackoverflow.com/search?q=pyav)
- [X] Looked through [old GitHub issues](https://github.com/PyAV-Org/PyAV/issues?&q=is%3Aissue)

## Additional context
If there's another way to specify this, then I'm definitely up for switching how I specify this, but I believe this is a valid use case since it's an encoder rather than a decoder and maybe the issue is that something is wrong with the `is_decoder` check?

**Comment by WyattBlue** [comment]
Can't repo. "done" is printed successfully.

```python
import av
from fractions import Fraction

container = av.open("test.mp4", "w")

container.add_stream("h264", time_base=Fraction(1, 1000))

container.close()
print("done")
```

**Comment by WyattBlue** [comment]
Your issue might be caused by compiling with Debian's old system libraries.

**Comment by daveisfera** [comment]
As I dug in more, it's a little more complicated than I realized because I had put in a workaround that I forgot about. I'm using `av.Codec("h264", "r")` to force it to be a decoder because the encoder version doesn't work. The first issue is that it doesn't allow you to set `extradata` (which I submitted a [PR to allow](https://github.com/PyAV-Org/PyAV/pull/1526)) but then it still doesn't result in a valid output even though the workaround I put in does. I'm guessing that I'm doing something wrong or there's a sneaky bug with some value not being set since I'm not just copying a codec from a template.

So running this with no second arg so `av.Codec("h264", "r")` will be used, results in a working file but adding `h264` as the second arg so it uses an encoder codec makes it not work.

```
import sys

import av

input_ = av.open(sys.argv[1])
output = av.open("output.mp4", "w")

# Make an output stream using the input as a template. This copies the stream
# setup from one to the other.
in_stream = input_.streams.video[0]
try:
    codec = sys.argv[2]
except IndexError:
    codec = av.Codec("h264", "r")
out_stream = output.add_stream(
    codec, width=in_stream.width, height=in_stream.height, rate=in_stream.average_rate, time_base=in_stream.time_base
)
out_stream.extradata = in_stream.extradata

for packet in input_.demux(in_stream):
    # print(packet)

    # We need to skip the "flushing" packets that `demux` generates.
    if packet.dts is None:
        continue

    # We need to assign the packet to the new stream.
    packet.stream = out_stream

    output.mux(packet)

input_.close()
output.close()
```

**Comment by daveisfera** [comment]
I noticed that the pixel format is missing when it uses the encoder, so is it doing something to set that when the decoder is used? Is there a way for me to ensure all of the codec values are set properly?

**Comment by daveisfera** [comment]
To check that theory, I did `print(name, getattr(out_stream.codec_context, name))` for each of the `out_stream` values with the different `codec` inputs and they were basically the same (the only difference was the GOP, name and the encoder/decoder flags), so dug some more and what's happening is the `extradata` on `out_stream` is changing the first time `output.mux()` is being called and that's why nothing ends up being written

**Reply by daveisfera** [reply]
The issue is that `template=in_stream` sets the codec to `h264` as a decoder, so the init of that codec is called and that allows everything to happen as expected. When `h264` is used directly, the PyAV interface sets `mode` to `w`, so it's an encoder (as it should be), but that means it's actually `libx264` and the init for that is called, which overwrites `extradata` and causes the problems I'm seeing here.
So long story short, the PR that I submitted allows this to all work correctly and it should be good now.
